### PR TITLE
#21/testing

### LINKED ISSUE #21: testing on the frontend 🤔
### Expected Behavior

There should be automated tests checking the behaviour of the frontend

- do flows get triggered at the expected time / with the expected behaviour
- do users see the layout we expect and get the response we planned when interacting?

### Current Behavior

n/a

### Any other notes

while we would usually apply end-to-end testing for this, developing an extension is a bit trickier in this regard so one possiblity is to invest in robust unit testing for all our utils, combined with the testing of endpoints on the backend repo

- this could be done with `jest`

we can also try using integration testing of our components

- this could be done with `cypress` or `storybook`

### Assignment

This issue is free for anyone to take

### PR DESCRIPTION
**Pull request by camelPhonso**
# Description

**Closes #21**  

Sets up jest for unit testing and creates an example file testing the handler for `updateMessage` method on our context.

### Files changed

- `jest.config.mjs`, `tsconfig.json`, `pnpm-lock.yaml`, `package.json` - installed new dependencies and updated configurations to allow jest to play nicely with our codebase
- `.husky/pre-commit` - added a check to our pre-commit hook that runs any tests associated with any files we have edited in that commit
- `background/index.ts` - renamed `background.ts` in accordance with @plasmo/messaging documentation
- `set_context.test.ts` - example unit test for our `set_context` handler

### UI changes

n/a

### Changes to Documentation

 - `README.md` and `TESTING.md` - updated with guidance on tests

# Tests

```bash
pnpm test // for all tests

pnpm test:changed // for any tests associated with files edited since the last commit

```

### CODE REVIEW

**Comment by camelPhonso** [comment]
Replaced the example test with a test for the new store method after merging dev ✨

File: src/background/messages/fetchCurrentProject.test.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import type { PlasmoMessaging } from "@plasmohq/messaging"
+
+import * as testData from "../../../test-data.json"
+import handler from "./fetchCurrentProject"
+
+fetchMock.enableMocks()
+
+describe("fetchCurrentProject", () => {
+  beforeEach(() => {
```
Comment (nichgalzin) [inline src/background/messages/fetchCurrentProject.test.ts]: My IDE is saying that the import call is missing for beforeEach, which is strange as the tests still pass without it. But it stops complaining when you add it the the @jest/globals import statement.

**Review by nichgalzin — CHANGES_REQUESTED** [review]
Mind removing the updateMessage utils? Or if we want to fix it make it a generic message handler function perhaps? Any, it has a type error that showing in the problems log. Thank youuuuuuu!

File: src/background/messages/fetchCurrentProject.test.ts (hunk 1 above)
Comment (camelPhonso) [reply src/background/messages/fetchCurrentProject.test.ts]: added that import now 👍

**Review by nichgalzin — APPROVED** [review]
Great stuff

**Pull request merged**